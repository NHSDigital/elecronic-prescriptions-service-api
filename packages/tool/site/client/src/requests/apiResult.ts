import {FhirResource} from "fhir/r4"

export function isApiResult(data: unknown): data is ApiResult {
  const result = data as ApiResult
  return typeof result.success === "boolean"
    && "request" in result
    && "response" in result
}

export interface ApiResult {
  success: boolean
  request: FhirResource
  request_xml?: string
  response: FhirResource
  response_xml?: string
}
