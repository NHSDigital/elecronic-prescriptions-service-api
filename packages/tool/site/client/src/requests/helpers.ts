export interface HubResponse {
  certificate: string,
  status_code: number,
  status_string: string,
  message: string,
  timestamp: string,
  signatures: Array<Signature>,
  failed_signatures: Array<Signature>
}

export interface Signature {
  id: string,
  signature: string
}

export function getToken(queryString: URLSearchParams): string {
  const token = queryString.get("token")

  if (!token) {
    throw new TypeError("Token not present in query string.")
  }

  return token
}

export function getState(queryString: URLSearchParams): string | null {
  return queryString.get("state")
}

export function getQueryString(): URLSearchParams {
  const queryString = window.location.search
  if (!queryString) {
    throw new TypeError("Empty query string.")
  }

  return new URLSearchParams(queryString)
}

export async function start(signFn: (jwt: string) => Promise<HubResponse>): Promise<void> {
  try {
    //showStatus("Signing in progress", "Downloading the signature request from the Signing Service.")
    const queryString = getQueryString()
    const token = getToken(queryString)
    const signatureRequest = await retrieveSignatureRequestFromServer(token)

    //showStatus("Signing in progress", "Communicating with Credential Management to generate a signature.")
    const hubResponse = await signFn(signatureRequest)
    validateHubResponse(hubResponse)

    //showStatus("Signing in progress", "Uploading the signature to the Signing Service.")
    const redirectUrl = await sendResponseToProvider(hubResponse, token)

    //showStatus("Signing complete", "Sending you back to the calling application.")
    const state = getState(queryString)
    window.location.href = state ? `${redirectUrl}&state=${state}` : redirectUrl
  } catch (e) {
    console.log(e)
    showError(e.message)
  }
}

function validateHubResponse(hubResponse: HubResponse) {
  if (!hubResponse) {
    throw new TypeError("No response from Credential Management.")
  }

  if (hubResponse.status_code !== 0) {
    console.log(`Got error response from Credential Management: ${JSON.stringify(hubResponse)}`)
    throw new TypeError("Error response from Credential Management.")
  }

  if (!hubResponse.certificate || !hubResponse.signatures?.length) {
    console.log(`Got invalid response from Credential Management: ${JSON.stringify(hubResponse)}`)
    throw new TypeError("Invalid response from Credential Management.")
  }
}

async function retrieveSignatureRequestFromServer(token: string): Promise<string> {
  const response = await fetch(`./client/signaturerequest/${token}`, {
    headers: {
      "X-CSRF-Token": getCSRFTokenFromCookie()
    }
  })
  if (response.status !== 200) {
    console.error(`Got unexpected response from Signing Service: ${response.status}`)
    throw new Error("Failed to retrieve signature request from Signing Service.")
  }
  const responseText = await response.text()
  if (!responseText) {
    throw new Error("Failed to parse response from Signing Service.")
  }
  return responseText
}

function getCSRFTokenFromCookie(): string {
  const cookies = getCookies()
  const crumb = cookies.get("crumb")

  if (!crumb) {
    throw new Error("Could not find CSRF token")
  }

  return crumb
}

function getCookies(): Map<string, string> {
  try {
    return new Map(document.cookie.split(";").map(cookie => {
      const cookieKey = cookie.split("=")[0].trim()
      const cookieValue = cookie.split("=")[1].trim()

      return [cookieKey, cookieValue]
    }))
  } catch(e) {
    console.error("Could not read cookies")
    console.error(e)
    return new Map()
  }
}

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const element = (identifier: string) => document.getElementById(identifier)!

export function showError(details?: string): void {
  element("status-container").style.display = "none"
  if (details) {
    element("error-details").innerText = `Error details: ${details}`
  }
  element("error-container").style.display = "block"
}

export async function sendResponseToProvider(hubResponse: HubResponse, token: string): Promise<string> {
  const response = await fetch(
    `./client/signatureresponse/${token}`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-CSRF-Token": getCSRFTokenFromCookie()
      },
      body: JSON.stringify({
        "signatures": hubResponse.signatures,
        "certificate": hubResponse.certificate
      })
    }
  )

  if (response.status !== 200) {
    console.error(`Got unexpected response from Signing Service: ${response.status}`)
    throw new Error("Failed to upload signature to Signing Service.")
  }

  const responseJson = await response.json()
  if (!responseJson) {
    throw new Error("Failed to parse response from Signing Service.")
  }

  return responseJson.redirectUri
}
