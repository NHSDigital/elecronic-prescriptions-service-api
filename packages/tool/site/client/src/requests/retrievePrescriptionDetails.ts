/* eslint-disable max-len */
import {axiosInstance} from "./axiosInstance"
import * as fhir from "fhir/r4"
import {getResponseDataIfValid} from "./getValidResponse"
import {getArrayTypeGuard, isBundle} from "../fhir/typeGuards"

export async function getPrescriptionOrderMessage(baseUrl: string, prescriptionId: string): Promise<fhir.Bundle> {
  const prescriptionOrderResponse = await axiosInstance.get<fhir.Bundle>(`${baseUrl}dispense/release/${prescriptionId}`)
  return getResponseDataIfValid(prescriptionOrderResponse, isBundle)
}

export async function getDispenseNotificationMessages(baseUrl: string, prescriptionId: string): Promise<Array<fhir.Bundle>> {
  const dispenseNotificationsResponse = await axiosInstance.get<Array<fhir.Bundle>>(`${baseUrl}dispenseNotifications/${prescriptionId}`)
  return getResponseDataIfValid(dispenseNotificationsResponse, getArrayTypeGuard(isBundle))
}
