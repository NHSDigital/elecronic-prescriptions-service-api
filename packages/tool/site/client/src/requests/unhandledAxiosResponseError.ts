import {AxiosResponse} from "axios"

export class UnhandledAxiosResponseError extends Error {
  response: AxiosResponse

  constructor(message: string, response: AxiosResponse) {
    super(message)
    this.response = response
  }
}
