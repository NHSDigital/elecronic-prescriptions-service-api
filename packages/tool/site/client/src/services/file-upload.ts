import {Dispatch, SetStateAction} from "react"

export function readPrescriptionsFromFiles(
  // eslint-disable-next-line no-undef
  files: FileList,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  prescriptionFilesUploaded: Array<any>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  setPrescriptionFilesUploaded: Dispatch<SetStateAction<Array<any>>>
): void {
  for (let i = 0; i < files.length; i++) {
    // eslint-disable-next-line no-undef
    const reader = new FileReader()
    reader.onload = event => {
      const result = event.target.result
      prescriptionFilesUploaded.push(result)
      setPrescriptionFilesUploaded(prescriptionFilesUploaded)
    }
    reader.readAsText(files[i])
  }
}
