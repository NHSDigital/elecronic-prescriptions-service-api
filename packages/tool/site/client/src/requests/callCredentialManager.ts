import {HubResponse, start} from "./helpers"

// eslint-disable-next-line @typescript-eslint/no-explicit-any
(window as any).jQuery = jQuery;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
(window as any).$ = jQuery

const SIGNALR_URL = `"http://localhost:"+prService.portNumber()+"/signalr"`
const SIGNALR_HUB_NAME = "signingHub"
const SIGNALR_METHOD_NAME = "requestToSign"

export async function sign(jwt: string): Promise<HubResponse> {
  const connection = $.hubConnection(SIGNALR_URL, {logging: true})
  if (connection.state !== SignalR.ConnectionState.Disconnected) {
    console.log(`Unexpected SignalR connection state ${connection.state}`)
    throw new Error("SignalR connection is already active.")
  }

  try {
    const proxy = connection.createHubProxy(SIGNALR_HUB_NAME)
    await connection.start({transport: "longPolling"})
    return await proxy.invoke(SIGNALR_METHOD_NAME, jwt)
  } finally {
    connection.stop()
  }
}

window.addEventListener("DOMContentLoaded", () => {
  start(sign)
})
