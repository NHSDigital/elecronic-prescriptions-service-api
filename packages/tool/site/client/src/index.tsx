import * as React from "react"
import {Page} from "./components/page"
import SignPage from "./pages/signPage"
import * as ReactDOM from "react-dom"
import {BrowserRouter, Route, Switch} from "react-router-dom"
import ClaimPage from "./pages/claimPage"
import DispensePage from "./pages/dispensePage"
import {CookiesProvider} from "react-cookie"
import SendPage from "./pages/sendPage"
import HomePage from "./pages/homePage"
import PrescriptionSearchPage from "./pages/prescriptionSearchPage"
import ReleasePage from "./pages/releasePage"
import CancelPage from "./pages/cancelPage"
import MyPrescriptionsPage from "./pages/myPrescriptionsPage"
import LoginPage from "./pages/loginPage"
import ValidatePage from "./pages/validatePage"
import ReturnPage from "./pages/returnPage"
import WithdrawPage from "./pages/withdrawPage"
import LoadPage from "./pages/loadPage"
import {Environment} from "./services/environment"
import LogoutPage from "./pages/logoutPage"
import ConfigPage from "./pages/configPage"
import ComparePage from "./pages/comparePage"
import ViewPrescriptionPage from "./pages/viewPrescriptionPage"
import {PageFullWidth} from "./components/pageFullWidth"
import {PageContainer} from "./components/pageContainer"
import DoseToTextPage from "./pages/doseToTextPage"
import TrackerViewPrescriptionPage from "./pages/trackerViewPrescription"

const customWindow = window as Record<string, any>

export interface AppContextValue {
  baseUrl: string
  environment: Environment
}

export const AppContext = React.createContext<AppContextValue>({baseUrl: "/", environment: "int"})

async function startApplication(baseUrl: string, environment: Environment): Promise<void> {
  const urlParams = new URLSearchParams(window.location.search)
  const content = (
    <AppContext.Provider value={{baseUrl, environment}}>
      <CookiesProvider>
        <Page>
          <BrowserRouter>
            <Switch>
              <Route path={`${baseUrl}/`}>
                <PageContainer>
                  <HomePage />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dose-to-text`}>
                <PageContainer>
                  <DoseToTextPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}config`}>
                <PageContainer>
                  <ConfigPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}my-prescriptions`}>
                <PageContainer>
                  <MyPrescriptionsPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}login`}>
                <PageContainer>
                  <LoginPage separateAuth={urlParams.get("separate_auth")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}logout`}>
                <PageContainer>
                  <LogoutPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/load`}>
                <PageContainer>
                  <LoadPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/edit`}>
                <PageContainer>
                  <SignPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/send`}>
                <PageContainer>
                  <SendPage token={urlParams.get("token")} state={urlParams.get("state")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/cancel`}>
                <PageContainer>
                  <CancelPage prescriptionId={urlParams.get("prescription_id")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/release`}>
                <PageContainer>
                  <ReleasePage prescriptionId={urlParams.get("prescription_id")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/return`}>
                <PageContainer>
                  <ReturnPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/withdraw`}>
                <PageContainer>
                  <WithdrawPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/dispense`}>
                <PageContainer>
                  <DispensePage prescriptionId={urlParams.get("prescription_id")} amendId={urlParams.get("amend_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/claim`}>
                <PageContainer>
                  <ClaimPage prescriptionId={urlParams.get("prescription_id")} amend={!!urlParams.get("amend")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}search`}>
                <PageContainer>
                  <PrescriptionSearchPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}tracker`}>
                <PageContainer>
                  <TrackerViewPrescriptionPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}view`}>
                <PageContainer>
                  <ViewPrescriptionPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}validate`}>
                <PageContainer>
                  <ValidatePage />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}compare-prescriptions`}>
                <PageFullWidth>
                  <ComparePage />
                </PageFullWidth>
              </Route>
            </Switch>
          </BrowserRouter>
        </Page>
      </CookiesProvider>
    </AppContext.Provider>
  )
  ReactDOM.render(content, document.getElementById("root"))
}

customWindow.startApplication = startApplication
