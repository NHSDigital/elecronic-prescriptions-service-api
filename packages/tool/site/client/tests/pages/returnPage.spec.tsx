import {waitFor} from "@testing-library/react"
import {screen} from "@testing-library/dom"
import pretty from "pretty"
import * as React from "react"
import MockAdapter from "axios-mock-adapter"
import {AppContextValue} from "../../src"
import {renderWithContext} from "../renderWithContext"
import userEvent from "@testing-library/user-event"
import {axiosInstance} from "../../src/requests/axiosInstance"
import ReturnPage from "../../src/pages/returnPage"
import {internalDev} from "../../src/services/environment"
import {MemoryRouter} from "react-router"

const baseUrl = "baseUrl/"
const prescriptionId = "7A9089-A83008-56A03J"
const context: AppContextValue = {baseUrl, environment: internalDev}

const returnUrl = `${baseUrl}dispense/return`

const mock = new MockAdapter(axiosInstance)

beforeEach(() => mock.reset())
afterEach(() => mock.reset())

test("Displays return form", async () => {
  const container = await renderPage()

  expect(screen.getByText("Return prescription")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays return result", async () => {
  mock.onAny(returnUrl).reply(200, {
    prescriptionIds: [],
    success: true,
    request: {req: "JSON Request"},
    request_xml: "XML Request",
    response: {res: "JSON Response"},
    response_xml: "XML Response"
  })

  const container = await renderPage()
  userEvent.click(screen.getByText("Return"))
  await waitFor(() => screen.getByText("Sending return."))
  await waitFor(() => screen.getByText(/Return Result/))
  expect(screen.getByText((/JSON Request/))).toBeTruthy()
  expect(screen.getByText("XML Request")).toBeTruthy()
  expect(screen.getByText((/JSON Response/))).toBeTruthy()
  expect(screen.getByText("XML Response")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

async function renderPage() {
  const {container} = renderWithContext(<MemoryRouter><ReturnPage prescriptionId={prescriptionId}/></MemoryRouter>, context)
  await waitFor(() => screen.getByText("Return prescription"))
  return container
}
