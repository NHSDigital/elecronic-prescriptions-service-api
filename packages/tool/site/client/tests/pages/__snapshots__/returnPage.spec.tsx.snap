// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays return form 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Return prescription</label></h1>
<dl class="nhsuk-summary-list">
  <div class="nhsuk-summary-list__row">
    <dt class="nhsuk-summary-list__key">ID</dt>
    <dd class="nhsuk-summary-list__value">7A9089-A83008-56A03J</dd>
  </div>
</dl>
<form>
  <div class="nhsuk-form-group">
    <fieldset class="nhsuk-fieldset">
      <div class="nhsuk-form-group"><label class="nhsuk-label nhsuk-label--s" id="pharmacy--label" for="pharmacy">Pharmacy returning prescription</label>
        <div class="nhsuk-radios" id="pharmacy" aria-labelledby="pharmacy--label" value="VNFKT">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-1" name="pharmacy" aria-labelledby="pharmacy-1--label" type="radio" value="VNFKT" checked=""><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-1--label" for="pharmacy-1">VNFKT - FIVE STAR HOMECARE LEEDS LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-2" name="pharmacy" aria-labelledby="pharmacy-2--label" type="radio" value="YGM1E"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-2--label" for="pharmacy-2">YGM1E - MBBM HEALTHCARE TECHNOLOGIES LTD</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="pharmacy-3" name="pharmacy" aria-labelledby="pharmacy-3--label" type="radio" value="custom"><label class="nhsuk-label nhsuk-radios__label" id="pharmacy-3--label" for="pharmacy-3">Other</label></div>
        </div>
      </div>
      <div class="nhsuk-form-group"><label class="nhsuk-label nhsuk-label--s" id="reason--label" for="reason">Choose a reason for returning</label>
        <div class="nhsuk-radios" id="reason" aria-labelledby="reason--label" value="0002">
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="reason-1" name="reason" aria-labelledby="reason-1--label" type="radio" value="0001"><label class="nhsuk-label nhsuk-radios__label" id="reason-1--label" for="reason-1">Patient non-attendance</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="reason-2" name="reason" aria-labelledby="reason-2--label" type="radio" value="0002" checked=""><label class="nhsuk-label nhsuk-radios__label" id="reason-2--label" for="reason-2">Unable to dispense medication on prescriptions</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="reason-3" name="reason" aria-labelledby="reason-3--label" type="radio" value="0003"><label class="nhsuk-label nhsuk-radios__label" id="reason-3--label" for="reason-3">Patient requested release</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="reason-4" name="reason" aria-labelledby="reason-4--label" type="radio" value="0004"><label class="nhsuk-label nhsuk-radios__label" id="reason-4--label" for="reason-4">Another dispenser requested release on behalf of the patient</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="reason-5" name="reason" aria-labelledby="reason-5--label" type="radio" value="0005"><label class="nhsuk-label nhsuk-radios__label" id="reason-5--label" for="reason-5">Invalid digital signature</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="reason-6" name="reason" aria-labelledby="reason-6--label" type="radio" value="0006"><label class="nhsuk-label nhsuk-radios__label" id="reason-6--label" for="reason-6">Rejected due to version problem</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="reason-7" name="reason" aria-labelledby="reason-7--label" type="radio" value="0007"><label class="nhsuk-label nhsuk-radios__label" id="reason-7--label" for="reason-7">Prescription otherwise invalid or unreadable</label></div>
          <div class="nhsuk-radios__item"><input class="nhsuk-radios__input" id="reason-8" name="reason" aria-labelledby="reason-8--label" type="radio" value="0008"><label class="nhsuk-label nhsuk-radios__label" id="reason-8--label" for="reason-8">Prescription expired</label></div>
        </div>
      </div>
    </fieldset>
  </div>
  <div class="sc-dlfmHC kSteAI"><button class="nhsuk-button" aria-disabled="false" type="submit">Return</button><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>
</form>"
`;

exports[`Displays return result 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Return Result <svg class="nhsuk-icon nhsuk-icon__tick" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" aria-hidden="true" height="34" width="34">
      <path stroke-width="4" stroke-linecap="round" stroke="#007f3b" fill="none" d="M18.4 7.8l-8.5 8.4L5.6 12"></path>
    </svg></label></h1>
<div class="nhsuk-expander-group">
  <details class="nhsuk-details nhsuk-expander">
    <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">1 - FHIR Request that a supplier would send to the FHIR API</span></summary>
    <div class="nhsuk-details__text">
      <div class="sc-dlfmHC kSteAI"><button class="nhsuk-button sc-jSguLX nyBIK" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jSguLX nyBIK" role="button" aria-disabled="false" draggable="false" href="data:application/json;charset=utf-8,%7B%0A%20%20%22req%22%3A%20%22JSON%20Request%22%0A%7D" download="message">Download</a></div>
      <div class="react-json-view" style="font-family: monospace; cursor: default; background-color: rgb(255, 255, 255); position: relative; margin-top: 10px;">
        <div class="pretty-json-container object-container">
          <div class="object-content">
            <div class="object-key-val"><span><span style="display: inline-block; cursor: pointer;"><div class="icon-container" style="display: inline-block; width: 17px;"><span class="expanded-icon"><svg style="vertical-align: top; color: rgb(0, 0, 0); height: 1em; width: 1em; padding-left: 2px;" viewBox="0 0 15 15" fill="currentColor"><path d="M0 5l6 6 6-6z"></path></svg></span></div><span></span><span style="display: inline-block; cursor: pointer; font-weight: bold; color: rgb(0, 94, 184);">{</span></span>
            <div style="display: inline-block; padding: 0px 0px 0px 10px;" class="object-meta-data"><span class="copy-to-clipboard-container" title="Copy to clipboard" style="vertical-align: top; display: none;"><span style="cursor: pointer; display: inline;"><span class="copy-icon"><svg style="vertical-align: top; color: rgb(0, 0, 0); font-size: 15px; margin-right: 3px; height: 1em; width: 1em;" viewBox="0 0 40 40" fill="currentColor" preserveAspectRatio="xMidYMid meet"><g><path d="m30 35h-25v-22.5h25v7.5h2.5v-12.5c0-1.4-1.1-2.5-2.5-2.5h-7.5c0-2.8-2.2-5-5-5s-5 2.2-5 5h-7.5c-1.4 0-2.5 1.1-2.5 2.5v27.5c0 1.4 1.1 2.5 2.5 2.5h25c1.4 0 2.5-1.1 2.5-2.5v-5h-2.5v5z m-20-27.5h2.5s2.5-1.1 2.5-2.5 1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5 1.3 2.5 2.5 2.5h2.5s2.5 1.1 2.5 2.5h-20c0-1.5 1.1-2.5 2.5-2.5z m-2.5 20h5v-2.5h-5v2.5z m17.5-5v-5l-10 7.5 10 7.5v-5h12.5v-5h-12.5z m-17.5 10h7.5v-2.5h-7.5v2.5z m12.5-17.5h-12.5v2.5h12.5v-2.5z m-7.5 5h-5v2.5h5v-2.5z"></path></g></svg></span></span></span></div></span>
            <div class="pushed-content object-container">
              <div class="object-content" style="margin-left: 6px;">
                <div style="border-left: 1px solid #000000; padding: 3px 5px 3px 20px;" class="variable-row"><span><span style="display: inline-block; color: rgb(0, 94, 184); letter-spacing: 0.5px; font-style: none; vertical-align: top; opacity: 0.85;" class="object-key"><span style="vertical-align: top;">"</span><span style="display: inline-block;">req</span><span style="vertical-align: top;">"</span></span><span style="display: inline-block; margin: 0px 5px; color: rgb(0, 94, 184); vertical-align: top;">:</span></span>
                  <div class="variable-value" style="display: inline-block; padding-right: 6px; position: relative; cursor: default;">
                    <div style="display: inline-block; color: rgb(0, 0, 0);"><span class="string-value" style="cursor: default;">"JSON Request"</span></div>
                  </div><span class="copy-to-clipboard-container" title="Copy to clipboard" style="vertical-align: top; display: none;"><span style="cursor: pointer; display: inline;"><span class="copy-icon"><svg style="vertical-align: top; color: rgb(0, 0, 0); font-size: 15px; margin-right: 3px; height: 1em; width: 1em;" viewBox="0 0 40 40" fill="currentColor" preserveAspectRatio="xMidYMid meet"><g><path d="m30 35h-25v-22.5h25v7.5h2.5v-12.5c0-1.4-1.1-2.5-2.5-2.5h-7.5c0-2.8-2.2-5-5-5s-5 2.2-5 5h-7.5c-1.4 0-2.5 1.1-2.5 2.5v27.5c0 1.4 1.1 2.5 2.5 2.5h25c1.4 0 2.5-1.1 2.5-2.5v-5h-2.5v5z m-20-27.5h2.5s2.5-1.1 2.5-2.5 1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5 1.3 2.5 2.5 2.5h2.5s2.5 1.1 2.5 2.5h-20c0-1.5 1.1-2.5 2.5-2.5z m-2.5 20h5v-2.5h-5v2.5z m17.5-5v-5l-10 7.5 10 7.5v-5h12.5v-5h-12.5z m-17.5 10h7.5v-2.5h-7.5v2.5z m12.5-17.5h-12.5v2.5h12.5v-2.5z m-7.5 5h-5v2.5h5v-2.5z"></path></g></svg></span></span></span>
                </div>
              </div>
            </div><span class="brace-row"><span style="display: inline-block; cursor: pointer; font-weight: bold; color: rgb(0, 94, 184); padding-left: 3px;">}</span></span>
          </div>
        </div>
      </div>
    </div>
</div>
</details>
<details class="nhsuk-details nhsuk-expander">
  <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">2 - Equivalent hl7v3 that the FHIR API has sent to spine</span></summary>
  <div class="nhsuk-details__text">
    <div class="sc-dlfmHC kSteAI"><button class="nhsuk-button sc-jSguLX nyBIK" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jSguLX nyBIK" role="button" aria-disabled="false" draggable="false" href="data:text/xml;charset=utf-8,XML%20Request" download="message">Download</a></div><pre class="sc-eCstZk YATfn">XML Request</pre>
  </div>
</details>
<details class="nhsuk-details nhsuk-expander">
  <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">3 -  The hl7v3 spine response that the FHIR API received from spine</span></summary>
  <div class="nhsuk-details__text">
    <div class="sc-dlfmHC kSteAI"><button class="nhsuk-button sc-jSguLX nyBIK" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jSguLX nyBIK" role="button" aria-disabled="false" draggable="false" href="data:text/xml;charset=utf-8,XML%20Response" download="message">Download</a></div><pre class="sc-eCstZk YATfn">XML Response</pre>
  </div>
</details>
<details class="nhsuk-details nhsuk-expander">
  <summary class="nhsuk-details__summary"><span class="nhsuk-details__summary-text">4 - FHIR Response that a supplier will receive from the FHIR API</span></summary>
  <div class="nhsuk-details__text">
    <div class="sc-dlfmHC kSteAI"><button class="nhsuk-button sc-jSguLX nyBIK" aria-disabled="false" type="submit">Copy</button><a class="nhsuk-button sc-jSguLX nyBIK" role="button" aria-disabled="false" draggable="false" href="data:application/json;charset=utf-8,%7B%0A%20%20%22res%22%3A%20%22JSON%20Response%22%0A%7D" download="message">Download</a></div>
    <div class="react-json-view" style="font-family: monospace; cursor: default; background-color: rgb(255, 255, 255); position: relative; margin-top: 10px;">
      <div class="pretty-json-container object-container">
        <div class="object-content">
          <div class="object-key-val"><span><span style="display: inline-block; cursor: pointer;"><div class="icon-container" style="display: inline-block; width: 17px;"><span class="expanded-icon"><svg style="vertical-align: top; color: rgb(0, 0, 0); height: 1em; width: 1em; padding-left: 2px;" viewBox="0 0 15 15" fill="currentColor"><path d="M0 5l6 6 6-6z"></path></svg></span></div><span></span><span style="display: inline-block; cursor: pointer; font-weight: bold; color: rgb(0, 94, 184);">{</span></span>
          <div style="display: inline-block; padding: 0px 0px 0px 10px;" class="object-meta-data"><span class="copy-to-clipboard-container" title="Copy to clipboard" style="vertical-align: top; display: none;"><span style="cursor: pointer; display: inline;"><span class="copy-icon"><svg style="vertical-align: top; color: rgb(0, 0, 0); font-size: 15px; margin-right: 3px; height: 1em; width: 1em;" viewBox="0 0 40 40" fill="currentColor" preserveAspectRatio="xMidYMid meet"><g><path d="m30 35h-25v-22.5h25v7.5h2.5v-12.5c0-1.4-1.1-2.5-2.5-2.5h-7.5c0-2.8-2.2-5-5-5s-5 2.2-5 5h-7.5c-1.4 0-2.5 1.1-2.5 2.5v27.5c0 1.4 1.1 2.5 2.5 2.5h25c1.4 0 2.5-1.1 2.5-2.5v-5h-2.5v5z m-20-27.5h2.5s2.5-1.1 2.5-2.5 1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5 1.3 2.5 2.5 2.5h2.5s2.5 1.1 2.5 2.5h-20c0-1.5 1.1-2.5 2.5-2.5z m-2.5 20h5v-2.5h-5v2.5z m17.5-5v-5l-10 7.5 10 7.5v-5h12.5v-5h-12.5z m-17.5 10h7.5v-2.5h-7.5v2.5z m12.5-17.5h-12.5v2.5h12.5v-2.5z m-7.5 5h-5v2.5h5v-2.5z"></path></g></svg></span></span></span></div></span>
          <div class="pushed-content object-container">
            <div class="object-content" style="margin-left: 6px;">
              <div style="border-left: 1px solid #000000; padding: 3px 5px 3px 20px;" class="variable-row"><span><span style="display: inline-block; color: rgb(0, 94, 184); letter-spacing: 0.5px; font-style: none; vertical-align: top; opacity: 0.85;" class="object-key"><span style="vertical-align: top;">"</span><span style="display: inline-block;">res</span><span style="vertical-align: top;">"</span></span><span style="display: inline-block; margin: 0px 5px; color: rgb(0, 94, 184); vertical-align: top;">:</span></span>
                <div class="variable-value" style="display: inline-block; padding-right: 6px; position: relative; cursor: default;">
                  <div style="display: inline-block; color: rgb(0, 0, 0);"><span class="string-value" style="cursor: default;">"JSON Response"</span></div>
                </div><span class="copy-to-clipboard-container" title="Copy to clipboard" style="vertical-align: top; display: none;"><span style="cursor: pointer; display: inline;"><span class="copy-icon"><svg style="vertical-align: top; color: rgb(0, 0, 0); font-size: 15px; margin-right: 3px; height: 1em; width: 1em;" viewBox="0 0 40 40" fill="currentColor" preserveAspectRatio="xMidYMid meet"><g><path d="m30 35h-25v-22.5h25v7.5h2.5v-12.5c0-1.4-1.1-2.5-2.5-2.5h-7.5c0-2.8-2.2-5-5-5s-5 2.2-5 5h-7.5c-1.4 0-2.5 1.1-2.5 2.5v27.5c0 1.4 1.1 2.5 2.5 2.5h25c1.4 0 2.5-1.1 2.5-2.5v-5h-2.5v5z m-20-27.5h2.5s2.5-1.1 2.5-2.5 1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5 1.3 2.5 2.5 2.5h2.5s2.5 1.1 2.5 2.5h-20c0-1.5 1.1-2.5 2.5-2.5z m-2.5 20h5v-2.5h-5v2.5z m17.5-5v-5l-10 7.5 10 7.5v-5h12.5v-5h-12.5z m-17.5 10h7.5v-2.5h-7.5v2.5z m12.5-17.5h-12.5v2.5h12.5v-2.5z m-7.5 5h-5v2.5h5v-2.5z"></path></g></svg></span></span></span>
              </div>
            </div>
          </div><span class="brace-row"><span style="display: inline-block; cursor: pointer; font-weight: bold; color: rgb(0, 94, 184); padding-left: 3px;">}</span></span>
        </div>
      </div>
    </div>
  </div>
  </div>
</details>
</div>
<div class="sc-dlfmHC kSteAI"><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>"
`;
